velox/functions/sparksql/aggregates/ModeAggregate.cpp

```diff
@@ -0,0 +1,585 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/functions/sparksql/aggregates/ModeAggregate.h"
+
+#include "folly/container/F14Map.h"
+
+#include "velox/exec/AddressableNonNullValueList.h"
+#include "velox/exec/SimpleAggregateAdapter.h"
+#include "velox/exec/Strings.h"
+
+using namespace facebook::velox::aggregate::prestosql;
+using namespace facebook::velox::exec;
+
+namespace facebook::velox::functions::aggregate::sparksql {
+
+namespace {
+
+struct UnknownValueHash {
+  std::size_t operator()(const UnknownValue& val) const noexcept {
+    return folly::hasher<UnknownValue>{}(val);
+  }
+};
+
+template <
+    typename T,
+    typename Hash = std::hash<T>,
+    typename EqualTo = std::equal_to<T>>
+class ModeAggregate {
+ public:
+  using InputType = Row<T>;
+
+  using IntermediateType = Map<T, int64_t>;
+
+  using OutputType = T;
+
+  struct AccumulatorType {
+    using ValueMap = folly::F14FastMap<
+        T,
+        int64_t,
+        Hash,
+        EqualTo,
+        AlignedStlAllocator<std::pair<const T, int64_t>, 16>>;
+
+    // A map of T -> count.
+    ValueMap values;
+
+    explicit AccumulatorType(HashStringAllocator* allocator)
+        : values{AlignedStlAllocator<std::pair<const T, int64_t>, 16>(
+              allocator)} {}
+    static constexpr bool is_fixed_size_ = false;
+
+    void addInput(HashStringAllocator* /*allocator*/, arg_type<T> data) {
+      values[data]++;
+    }
+
+    void combine(
+        HashStringAllocator* /*allocator*/,
+        arg_type<IntermediateType> other) {
+      for (const auto& [key, count] : other) {
+        if (count.has_value()) {
+          values[key] += count.value();
+        }
+      }
+    }
+
+    bool writeIntermediateResult(out_type<IntermediateType>& out) {
+      out.reserve(values.size());
+      for (const auto& [value, count] : values) {
+        out.emplace(value, count);
+      }
+      return true;
```

Comment:
I wonder what's the intermediate result if values are empty, null or an empty map. Spark seems to return empty buffer which is consistent with current implementation, but could you help confirm? Thanks.

Reply:
It should be empty vector, nothing should be written to the writer.
https://github.com/facebookincubator/velox/blob/8ab9d11c95fbfd93d6381f95238adceef5109052/velox/exec/SimpleAggregateAdapter.h#L287C8-L287C27